hashintel/hash
# H-990, H-1067: Use graph permissions in frontend, readonly org settings view

## 🌟 What is the purpose of this PR?

This PR updates the Node API and frontend so that the frontend can rely on the permissions from the graph when determining what UI to show users, rather than the placeholder checks implemented prior to proper authorization.

As part of this it introduces a readonly view to the org settings and member management pages, which were previously editable by anyone who could view them.

## 🔍 What does this change?



**Node API**
- Update GraphQL return types:
  - `Subgraph` becomes `SubgraphAndUserPermissions`
  - `Page` gets `userPermissions`
  - `Comment` gets `canUserEdit`
  - ideally we wouldn't treat these latter two differently but we're better off getting rid of the special treatment of them in the GraphQL API entirely rather than trying to do half of it now
- Remove old checks in API resolvers

**Frontend**
- use the new checks everywhere we need to check if something is editable
- Org settings
  - expose `[add|remove]AccountGroupMember` to FE and use when adding/removing members
  - readonly view for membership/settings page 

**Drive-bys**
- The `EntityTypeMismatchError` was wrong in various places, probably due to copy and paste errors

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?




This PR:


### 📜 Does this require a change to the docs?




The changes in this PR:


### 🕸️ Does this require a change to the Turbo Graph?




The changes in this PR:


## ⚠️ Known issues

- All org members can edit the org profile. This will be changed so that only org admins/owners can edit it.
- Sharing a page with another user won't work properly as it only currently shares the main `Page` entity, not the blocks it links to. Thus the page will crash for other users, as the block entities are missing.


## 🛡 What tests cover this?



- A couple of existing Playwright, not much coverage.

## ❓ How to test this?



1. Check that features for a single user still work as expected
2. Check that other users can / can't edit things as appropriate based on sharing settings

## 📹 Demo



**Sharing an entity with a user**

https://github.com/hashintel/hash/assets/37743469/7b81c8cb-ced4-4795-a8cc-ce6e9bc0f1c5

---

**Adding/removing an org member**


https://github.com/hashintel/hash/assets/37743469/68e2bb06-dd1d-458c-b3bb-f23118d38962
